### PR TITLE
WIP - OAuth plugin

### PR DESCRIPTION
**Pull request by liggitt**
Initial work on an OAuth plugin package, based on github.com/openshift/origin/tree/master/pkg/oauth
- [x] API objects to back an OAuth server implementation
- [x] Pull in github.com/RangelReale/osin as godep
- [x] Define osin-based OAuth server
- [x] Add interfaces for pluggable auth, login, challenge, and grant flows
- [x] Add implementations for auth
  - [x] OAuth access token authenticator.Token (validates tokens as unexpired OAuth tokens)
  - [x] session-based authenticator.Request (checks request session for user info)
  - [x] external URL basicauth authenticator.Password (validates password using a basicauth request to an external URL)
  - [x] grant approval UI
  - [x] simple login flow
- [x] Rework OAuth etcd object registries to use new generic registry
- [x] Add a plugin/cmd/oauth command
- [x] Include example of hooking up OAuth object rest endpoints, token authenticator, and OAuth /authorize and /token endpoints

To do:
- [ ] Decide whether to hook in implemented OAuth client stuff to the main API client, or leave it as a separate client object
- [ ] Decide how something like the /authorize and /token endpoints would be registered (can't go under the same protected mux as the API, or you couldn't access it without an API token, causing a paradox)
- [ ] Enforce OAuth API objects being non-namespaced

### CODE REVIEW

**Comment by liggitt** [comment]
Coming back to this... did the rework with the generic etcd registry. Sample wiring in `plugin/cmd/oauth/oauth.go`, temporarily used in `master.go` to illustrate usage.

Sample usage:
```

# create demo oauth client
curl -s http://localhost:8080/api/v1beta1/oauthClients -X POST -d '{"apiVersion":"v1beta1","id":"demo","secret":"my_secret","redirectURIs":["http://localhost:8081/", "http://localhost:9000/"]}' | json

# access in browser, log in, approve grant
http://localhost:8080/oauth/authorize?response_type=token&client_id=demo&scope=scopeA&redirect_uri=http%3A%2F%2Flocalhost%3A9000%2Findex.html

# Copy the oauth token from redirect response URL hash
TOKEN=Zjg1M2MwYmEtYTY1MC00MDMyLTg2MGQtYTZhMmYxODFiMjlm

# test using the token
HOST=`hostname`
curl -k https://$HOST:8443/api/v1beta1/pods
curl -k https://$HOST:8443/api/v1beta1/pods -H "Authorization: Bearer $TOKEN"
curl -k https://$HOST:8443/api/v1beta1/pods -H "Authorization: Bearer XYZ"
```

**Comment by roberthbailey** [comment]
Jordan, what's the status of this?

**Comment by liggitt** [comment]
@roberthbailey rebased and updated. This is pretty close to what we have in OpenShift now. Ready for someone to take a look if they feel like it. The remaining TODO items are related to how and whether this would be referenced by the main API client and apiserver http handlers.

**Comment by liggitt** [comment]
I tried to keep the commits logically separated... I can squash if that helps. The last commit is meant as more of an example of how it could be wired in. I expect that would be discarded or reworked depending on how the last couple TODO items are decided

**Comment by roberthbailey** [comment]
I'll take a look tomorrow.

**Comment by liggitt** [comment]
no idea what is going on with Travis... test-go passes locally, and I don't see any differences in the conversion test failure output

**Comment by piosz** [comment]
Is this PR active? If not please close it according to https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/devel/pull-requests.md

**Comment by liggitt** [comment]
Yes, though lagging behind some refactorings

File: plugin/pkg/oauth/client/fake_authorize_tokens.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+package client
+
+import (
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/labels"
+	oapi "github.com/GoogleCloudPlatform/kubernetes/plugin/pkg/oauth/api"
+)
+
+// FakePods implements PodsInterface. Meant to be embedded into a struct to get a default
```
Comment (erictune) [inline plugin/pkg/oauth/client/fake_authorize_tokens.go]: s/FakePods/FakeAuthorizeTokens/

**Comment by erictune** [comment]
I don't completely understand this, but I am increasingly convinced we need this in kubernetes.  

It is a lot of new types, and I wonder @smarterclayton if you think the new types in this PR beling in a separate API from the core API, and how hard it is to do that?

**Comment by smarterclayton** [comment]
We had originally talked about exposing them via a different endpoint, api group, and versioning strategy, maybe within the plugin package.  I think this would be an excellent feature to demonstrate doing that, similar to David's comments in the other thread.

**Pull request closed without merging**